File: examples/clj/src/com/rpl/agent/basic/stream_all_agent.clj
(ns com.rpl.agent.basic.stream-all-agent
  "Demonstrates subscribing to streaming chunks across multiple agent invocations.

  Features demonstrated:
  - stream-chunk!: Emit streaming data from nodes
  - agent-stream-all: Subscribe to streaming data from all invocations of a node
  - Multiple agent invocations with single subscription
  - Tracking chunks by invoke ID
  - Invoke ID mapping in streaming callbacks"
  (:require
   [com.rpl.agent-o-rama :as aor]
   [com.rpl.rama :as rama]
   [com.rpl.rama.test :as rtest]))

;;; Agent module demonstrating stream-all functionality
(aor/defagentmodule StreamAllAgentModule
  [topology]

  (->
   (aor/new-agent topology "StreamAllAgent")

    ;; Node that processes a task and streams progress
   (aor/node
    "process-task"
    nil
    (fn [agent-node {:keys [task-id items-to-process]}]
      (println (format "\nProcessing task %s with %d items" task-id items-to-process))

       ;; Stream progress as we process items
      (doseq [item-num (range items-to-process)]
        (Thread/sleep 50)

         ;; Stream progress update
        (aor/stream-chunk! agent-node
                           {:task-id     task-id
                            :item-number item-num
                            :status      "processing"})

        (println (format "Task %s: Processed item %d/%d"
                         task-id
                         (inc item-num)
                         items-to-process)))

       ;; Return final result
      (aor/result! agent-node
                   {:task-id      task-id
                    :status       "completed"
                    :total-items  items-to-process
                    :completed-at (System/currentTimeMillis)})))))

(defn -main
  "Run the stream-all agent example"
  [& _args]
  (with-open [ipc (rtest/create-ipc)]
    (rtest/launch-module! ipc StreamAllAgentModule {:tasks 1 :threads 1})

    (let [manager (aor/agent-manager ipc
                                     (rama/get-module-name
                                      StreamAllAgentModule))
          agent   (aor/agent-client manager "StreamAllAgent")]

      (println "Stream-All Agent Example:")
      (println "Demonstrating streaming across multiple agent invocations...")

       ;; Track chunks received by invoke ID
      (let [chunks-by-invoke (atom {})
            ;; Start two invocations
            invoke1 (aor/agent-initiate agent
                                        {:task-id          "task-1"
                                         :items-to-process 3})
            invoke2 (aor/agent-initiate agent
                                        {:task-id          "task-2"
                                         :items-to-process 4})]

        (println "\nStarted 2 agent invocations")
        (println "Subscribing to streaming from all invocations...")

         ;; Subscribe to streaming chunks from ALL invocations
        (aor/agent-stream-all
         agent
         invoke1
         "process-task"
         (fn [_invoke-id->all-chunks invoke-id->new-chunks _reset-invoke-ids _complete?]
           (doseq [[invoke-id new-chunks] invoke-id->new-chunks]
             (swap! chunks-by-invoke update invoke-id
                    (fn [existing]
                      (concat (or existing []) new-chunks)))
             (doseq [chunk new-chunks]
               (println (format "Received streaming chunk: Task=%s Item=%d [invoke-id=%s]"
                                (:task-id chunk)
                                (:item-number chunk)
                                invoke-id))))))

         ;; Wait for all invocations to complete
        (println "\nWaiting for all invocations to complete...")
        (let [result1 (aor/agent-result agent invoke1)
              result2 (aor/agent-result agent invoke2)]

          (println "\nFinal results:")
          (println (format "  %s: %d items processed"
                           (:task-id result1)
                           (:total-items result1)))
          (println (format "  %s: %d items processed"
                           (:task-id result2)
                           (:total-items result2)))

          (println "\nStreaming summary:")
          (println (format "  Total invocations tracked: %d"
                           (count @chunks-by-invoke)))
          (doseq [[invoke-id chunks] @chunks-by-invoke]
            (println (format "  Invoke %s: received %d streaming chunks"
                             invoke-id
                             (count chunks))))

          (println "\nNotice how:")
          (println "- agent-stream-all subscribes to ALL invocations of an agent")
          (println "- Both invocations were started before subscribing")
          (println "- Chunks are grouped and delivered by invoke ID")
          (println "- A single callback handles all invocations")
          (println "- Each invocation is tracked independently"))))))

(comment
  (-main))

File: examples/java/basic/src/main/java/com/rpl/agent/basic/StreamAllAgent.java
package com.rpl.agent.basic;

import com.rpl.agentorama.AgentClient;
import com.rpl.agentorama.AgentInvoke;
import com.rpl.agentorama.AgentManager;
import com.rpl.agentorama.AgentNode;
import com.rpl.agentorama.AgentTopology;
import com.rpl.agentorama.AgentModule;
import com.rpl.agentorama.ops.RamaVoidFunction2;
import com.rpl.rama.test.InProcessCluster;
import com.rpl.rama.test.LaunchConfig;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import java.util.concurrent.ConcurrentHashMap;

/**
 * Java example demonstrating subscribing to streaming chunks across multiple agent invocations.
 *
 * <p>Features demonstrated:
 *
 * <ul>
 *   <li>streamChunk: Emit streaming data from nodes
 *   <li>agent.streamAll: Subscribe to streaming data from all invocations of a node
 *   <li>Multiple agent invocations with single subscription
 *   <li>Tracking chunks by invoke ID
 *   <li>Invoke ID mapping in streaming callbacks
 * </ul>
 *
 * <p>All required classes are defined as nested classes within this single file for simplicity and
 * self-containment.
 */
public class StreamAllAgent {

  /** Agent Module demonstrating stream-all functionality. */
  public static class StreamAllAgentModule extends AgentModule {

    @Override
    protected void defineAgents(AgentTopology topology) {
      topology.newAgent("StreamAllAgent").node("process-task", null, new ProcessTaskFunction());
    }
  }

  /** Node function that processes a task and streams progress updates. */
  public static class ProcessTaskFunction
      implements RamaVoidFunction2<AgentNode, Map<String, Object>> {

    @Override
    public void invoke(AgentNode agentNode, Map<String, Object> request) {
      String taskId = (String) request.get("taskId");
      int itemsToProcess = (Integer) request.get("itemsToProcess");

      System.out.printf("%nProcessing task %s with %d items%n", taskId, itemsToProcess);

      // Stream progress as we process items
      for (int itemNum = 0; itemNum < itemsToProcess; itemNum++) {
        try {
          Thread.sleep(50);
        } catch (InterruptedException e) {
          Thread.currentThread().interrupt();
          return;
        }

        // Stream progress update
        Map<String, Object> chunk = new HashMap<>();
        chunk.put("taskId", taskId);
        chunk.put("itemNumber", itemNum);
        chunk.put("status", "processing");
        agentNode.streamChunk(chunk);

        System.out.printf("Task %s: Processed item %d/%d%n", taskId, itemNum + 1, itemsToProcess);
      }

      // Return final result
      Map<String, Object> result = new HashMap<>();
      result.put("taskId", taskId);
      result.put("status", "completed");
      result.put("totalItems", itemsToProcess);
      result.put("completedAt", System.currentTimeMillis());
      agentNode.result(result);
    }
  }

  public static void main(String[] args) throws Exception {
    System.out.println("Starting Stream-All Agent Example...");

    try (InProcessCluster ipc = InProcessCluster.create()) {
      // Launch the agent module
      StreamAllAgentModule module = new StreamAllAgentModule();
      ipc.launchModule(module, new LaunchConfig(1, 1));

      // Get agent manager and client
      String moduleName = module.getModuleName();
      AgentManager manager = AgentManager.create(ipc, moduleName);
      AgentClient agent = manager.getAgentClient("StreamAllAgent");

      System.out.println("Stream-All Agent Example:");
      System.out.println("Demonstrating streaming across multiple agent invocations...");

      // Track chunks received by invoke ID
      Map<Object, List<Map<String, Object>>> chunksByInvoke = new ConcurrentHashMap<>();

      // Start two invocations
      Map<String, Object> request1 = new HashMap<>();
      request1.put("taskId", "task-1");
      request1.put("itemsToProcess", 3);
      AgentInvoke invoke1 = agent.initiate(request1);

      Map<String, Object> request2 = new HashMap<>();
      request2.put("taskId", "task-2");
      request2.put("itemsToProcess", 4);
      AgentInvoke invoke2 = agent.initiate(request2);

      System.out.println("\nStarted 2 agent invocations");
      System.out.println("Subscribing to streaming from all invocations...");

      // Subscribe to streaming chunks from ALL invocations
      agent.streamAll(
          invoke1,
          "process-task",
          (allChunks, newChunks, resetInvokeIds, complete) -> {
            for (Map.Entry<UUID, List<Object>> entry : newChunks.entrySet()) {
              UUID invokeId = entry.getKey();
              List<Object> chunks = entry.getValue();

              for (Object chunkObj : chunks) {
                @SuppressWarnings("unchecked")
                Map<String, Object> chunk = (Map<String, Object>) chunkObj;

                chunksByInvoke.computeIfAbsent(invokeId, k -> new ArrayList<>()).add(chunk);

                System.out.printf(
                    "Received streaming chunk: Task=%s Item=%d [invoke-id=%s]%n",
                    chunk.get("taskId"), chunk.get("itemNumber"), invokeId);
              }
            }
          });

      // Wait for all invocations to complete
      System.out.println("\nWaiting for all invocations to complete...");
      @SuppressWarnings("unchecked")
      Map<String, Object> result1 = (Map<String, Object>) agent.result(invoke1);
      @SuppressWarnings("unchecked")
      Map<String, Object> result2 = (Map<String, Object>) agent.result(invoke2);

      System.out.println("\nFinal results:");
      System.out.printf(
          "  %s: %d items processed%n", result1.get("taskId"), result1.get("totalItems"));
      System.out.printf(
          "  %s: %d items processed%n", result2.get("taskId"), result2.get("totalItems"));

      System.out.println("\nStreaming summary:");
      System.out.printf("  Total invocations tracked: %d%n", chunksByInvoke.size());
      for (Map.Entry<Object, List<Map<String, Object>>> entry : chunksByInvoke.entrySet()) {
        System.out.printf(
            "  Invoke %s: received %d streaming chunks%n", entry.getKey(), entry.getValue().size());
      }

      System.out.println("\nNotice how:");
      System.out.println("- streamAll subscribes to ALL invocations of an agent");
      System.out.println("- Both invocations were started before subscribing");
      System.out.println("- Chunks are grouped and delivered by invoke ID");
      System.out.println("- A single callback handles all invocations");
      System.out.println("- Each invocation is tracked independently");
    }
  }
}
